README.md
# 自动化测试
疑问及建议可发送邮件至：mizhidajiangyou@163.com

[环境](#环境)

[接口自动化测试](#接口自动化测试)

[UI自动化测试](#UI自动化测试)

[接口性能自动化测试](#接口性能自动化测试)

[磁盘性能自动化测试](#磁盘性能自动化测试)

[PC端自动化测试](#PC端自动化测试)

[测试服务器](#测试服务器)

## 正常使用
* docker pull xxxx(准备中)

* 执行Shell/z-env.sh 以完成linux场景下的环境变量配置,windows场景下自行配置ZHOME系统变量

* 在TestCase_Pytest下创建test开头的Py文件

* 根据需要引用Common里的函数可实现ui、接口、性能等自动化测试，可参照Demo文件中的一些模板。

* 参照Demo的runner.py生成基于allure的测试报告

## 依赖
`pip install matplotlib`

`pip install pytest`

`pip install allure-pytest`

`git clone https://github.com/locustio/locust.git && python setup.py install`

1.5版本使用注意事项: https://docs.locust.io/en/latest/quickstart.html

Vue3.0 :`<script src="https://unpkg.com/vue@next"></script>`
ElementPlus:`<link rel="stylesheet" href="https://unpkg.com/element-plus/lib/theme-chalk/index.css">`
`<script src="https://unpkg.com/element-plus/lib/index.full.js"></script>`
* 可通过Tools/Web/downloadUnpkg.py文件将依赖下载至本地（默认将下载最新版本[-1]，默认存放Tools/Web路径（./））

`pip install Selenium`
Chrome：https://chromedriver.storage.googleapis.com/index.html?path=91.0.4472.19/
Firefox：https://github.com/mozilla/geckodriver/releases

`pip install pytest-shutil`

`pip install pymongo`

`pip install pymysql`

`pip install Django -i https://pypi.tuna.tsinghua.edu.cn/simple`

## 环境

### log支持中文
修改FileHandler => def __init__(self, filename, mode='a', encoding='utf-8', delay=False)
    
### 使用pytest
左上角file->Setting->Tools->Python Integrated Tools->项目名称->Default test runner->选择py.test

### 使用allure
1. 下载：https://repo.maven.apache.org/maven2/io/qameta/allure/allure-commandline/2.13.9/
2. path中添加

### linux解决SimHei问题
1、import matplotlib print(matplotlib.matplotlib_fname()) 获取路径；
2、拷贝至mpl-data/fonts/ttf 
3、删除缓存：rm ~/.cache/matplotlib -rf（注意：可能缓存地址不对，使用代码 fontmanager.get_cachedir() 或 fontmanager._fmcache获取缓存地址。） 
4、plt.rcParams['font.sans-serif']=['SimHei']  

### MongoDB
MongoDB：https://www.mongodb.com/try/download/community 
* linux：选择TGZ下载， 解压tar -zxvf mongodb-linux.tgz；改名为mongodb；;加入环境变量：export PATH=<mongodb-install-directory>/bin:$PATH；cd mongodb &&mkdir -p data/db；运行./mongod --port 27017 --dbpath=/data/mongodb/data/db --logpath=/data/mongodb/logs --fork;查看ps -ef
* Windows：下载msi安装；在data目录下创建db、log

## 接口自动化测试

## UI自动化测试

### 模块构成
Common、Report/UI、TestCase

### 核心模块

#### 元素操作模块
elementOperate.py

##### 功能说明
* 提供异常处理，重试次数，重试间隔在Config/currency中设置
* 提供css_selector、xpath、id三种方法搜索元素
* 封装元素click、sendkeys、clear等操作
* 提供异常情况下截图并存放UI/ErrorPng
* 提供全屏截图功能

## 接口性能自动化测试
使用Locust，模版放置Demo。 待封装待优化。
### 模块构成

## 磁盘性能自动化测试 
目前支持fio和vdbench2种工具自动化测试

### 模块构成
Performance、Shell/fio、Shell/vdbench、Report/DiskPerformance、Common、Demo/DiskPerformanceReport

### 核心模块

#### vdbench自动测试模块
Shell/vdbench/run-vdbench.sh
支持在linux任意目录运行该脚本，可多客户机运行。

##### 功能及参数说明
./run-vdbench.sh --help 查看帮助信息，以修改自动生成的vdbench测试脚本配置。
默认自动生成FC的vdbench脚本，并自动执行标准化的4K、1M，随机、顺序，读、写性能测试。

    <--brand| --mode| --type| --ip> \n
    (--Ldisk)\n
    [--size| --rdpct| --block| --fileio| --seekpct] \n
    [--runtime| --interval| --warmup | --pause] \n
    [--file| --out| --log| --date] \n
    brand     <string>            disk manufacturer;default SEAGATE
    mode      <int>               whether which run mode you test;default 2
              *   0               scan and test
              *   1               test and make report
              *   2               nohup run vdbench &
              *   3               only use your output to make picture report
              *   4               ssh no password
              *   5               use parameter \"command\" to run bash in all client
    type      <string>            whether which volume type you test;default fc
              *   fc              must install device-mapper-multipath(multipath-tools)
              *   iscsi           must install iscsi
              *   nfs             must install nfs-utils
              *   cifs            default support
              *   Ldisk           test every client local disk whitch you  appoint
              *   Lfile           test every client local file system  whitch you  appoint
    disk      <\"array\">           when the type is Ldisk must to define;no default.
    ip        <\"array\">           all ip list which you want to test;default ssh ip
    size      <int>               disk or file size;default 500
    rdpct     <\"array\">           percentage of read ;default \"0 100\"
    block     <\"array\">           test block size ;default \"4k 1M\"
    fileio    <\"array\">           test fileio ;default \"random sequential\"
    seekpct   <\"array\">           test random ratio ;default \"100 0\"
    runtime   <int>               test runtime(s) ;default 300
    interval  <int>               print interval(s) ;default 1
    warmup    <int>               hot start time(s) ;default 30
    pause     <int>               pause time(s) ;default 30
    fdepth    <int>               fsd default file depth
    fwidth    <int>               fsd default file width
    fnum      <int>               fsd default file num
    fsize     <int>               fsd default file size (MB)
    file      <\"path\">            *.vbd will put in;default pwd
    out       <\"path\">            vdbench out put will put in;default pwd/vd-output
    log       <\"path\">            run logs will put in;default same with file
    date      <date>              date for test ,like 220101;default date '+%y%m%d'
    command   <string>            the command in ssh \"ip\" bash \"command\"
    
#### FIO自动化测试模块
Shell/zfioPerformance.sh
仅支持在本目录运行，待优化项较多，目前仅支持单机运行。

##### 功能及参数说明

    $1 参数1位类型
    $2 参数2为名称
    $3 参数3设定执行方式
    $4 测试磁盘是需使用参数4表示盘符

#### 报告生成模块
Common/fileOperate.py、reOperate.py
目前仅支持fio报告

##### 功能说明
* 根据地址、本次测试数据，生成HTML、JS、CSS（基于ElementPlus）
* `待选择，2种方案`[每次报告的数据会生成一个json格式文件/直接读取json文件并作出修改] 然后替换mainDemo.js中的内容再去生成js文件
* 使用selenium对生成的HTML进行全屏截图，保存文件即为测试报告

#### 图片生成模块
Performance/bar.py
Performance/makelines.py
支持fio（利用sysstat获取实时io；）vdbench（工具自身统计实时io）2种工具测试场景下，精确到s的IO折线图生成与最终测试结果的柱状图、折线图生成。
部分效果图见Report/DiskPerformance/2021-05-26/
效果：
![Image text](https://github.com/mizhidajiangyou/myTest/blob/performance/Report/DiskPerformance/2021-05-26/linepng-IOPS-2021-05-21.png)

Common/elementOperate.py
效果：<img src="https://github.com/mizhidajiangyou/myTest/blob/performance/Report/DiskPerformance/2021-05-26/all.png" width="100" height="200" />

Performance/IOLines.py
![Image text](https://github.com/mizhidajiangyou/myTest/blob/performance/Report/DiskPerformance/2021-05-26/disk-write.png)
##### 柱状图

##### 折线图

## PC端自动化测试

### 模块构成

### 核心模块

#### 元素操作模块


##### 功能说明


## 测试服务器
使用的linux系统为Ubuntu20.04

### 搭建

#### 一、ubuntu20.04系统部署
* IP配置
`vim /etc/netplan/xxxxx.yaml`

添加：
   network:
  ethernets:
    enp7s0f0:
      addresses:
      - 10.18.18.23/24
      gateway4: 10.18.18.1
      
  version: 2
 
 `netplan apply`

* 源配置 
`deb https://mirrors.tuna.tsinghua.edu.cn/ubuntu/ focal main restricted universe multiverse`
`deb https://mirrors.tuna.tsinghua.edu.cn/ubuntu/ focal-security main restricted universe multiverse`
`deb https://mirrors.tuna.tsinghua.edu.cn/ubuntu/ focal-updates main restricted universe multiverse`
`deb https://mirrors.tuna.tsinghua.edu.cn/ubuntu/ focal-proposed main restricted universe multiverse`
`deb https://mirrors.tuna.tsinghua.edu.cn/ubuntu/ focal-backports main restricted universe multiverse`
`https://mirror.tuna.tsinghua.edu.cn/help/ubuntu/`
* ssh启用
一般只会安装客户端，所以需要安装服务端（）
`sudo apt install openssh-server`

-> LoginGraceTime 2m
-> PermitRootLogin yes
-> StrictModes yes

#### 二、Apache2部署

* 安装apache2 `apt-get install apache2`

* 启动服务 /etc/init.d/apache2 restart/start/stop && apache2ctl -k  restart/start/stop

* 修改html根目录：/etc/apache2/sites-available/000-default.conf中的 DocumentRoot 对应的目录

* 修改html文件对应：/etc/apache2/mods-available/dir.conf（可不做）

#### 三、Django部署
* python manage.py runserver

* django-admin startproject MySite

* 修改MySite/settings => TIME_ZONE = 'Asia/Shanghai';LANGUAGE_CODE = 'zh-hans' ;USE_TZ = False

* 在init文件中追加 `import pymysql;pymysql.install_as_MySQLdb()`

* mysql配置：grant all privileges on test1.* to 'django'@'localhost' identified by 'password';

* settings配置文件修改：`DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.mysql',
        'NAME': 'test1',
        'USER': 'django',
        'PASSWORD': 'password',
        'HOST': 'localhost',
        'PORT': '3306',
    }
}`

* `python manage.py migrate` # 创建表结构

*  ORM模型：app(django-admin startapp)>>models中配置：`类名<->表名` ; `实例<->记录` ;`属性<->字段` 追加`class Question(models.Model):
    question_text = models.CharField(max_length=200)`

* setting.py配置文件修改：`INSTALLED_APPS = [
    'polls'
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
]`

* `python manage.py makemigrations` # 模型变更

* `python manage.py migrate polls` # 根据模型创建表















CHANGELOG.md
# 更新说明
## 版本0.0.1
* 支持FIO自动化测试磁盘、绘制实时IO图、最大值折线图并自动化生成测试报告。
* 通用接口自动化测试函数
* 通用UI自动化测试函数
## 版本0.0.2
* 增加CHANGELOG
* 增加磁盘自动化测试时获取测试设备硬件信息
* 支持所有dev目录下磁盘设备自动化磁盘测试
## 版本0.0.3
* 支持使用多个客户端进行联机磁盘、文件系统性能自动化测试并生成结果
* 利用python虚拟化解决版本包冲突问题
## 版本0.0.4
* _支持多终端场景，mysql压力/性能自动化测试并生成结果_ 待实现中...
* 通用变量均置于$ZHOME/Config/global.ini下
* 重写了日志模块


# 缺陷及优化

### 已知缺陷
* ~~iops/带宽值小于4的情况下无法生成barfix 21-0428）~~
* ~~图片路径不能复原（fix 21-0428）~~
* ~~没有转换KB/s为MB/s(fix 21-0521)~~
* ~~特殊数值导致bar脚本无法复原（fix 21-0429）~~
* ~~iops带k情况下最大值取不到的问题（fix 21-0505）~~
* ~~根据文件生成折线图IOPS图存在取值错误（fix 21-0520）~~
* ~~linux下无SimHei问题(fix 21-0521)~~
* 利用封装组件zCharts做出来的图纵坐标无法显示

### 待优化/处理、计划
* Linux环境下支持中文
* 传参问题，以精简代码
* ~~图表生成~~
* ~~$3支持选择执行步骤~~
* ~~最大值生成~~
* bw、iops判断大小的优化
* ~~根据最大值max文件生成折线图~~
* 直接输出磁盘性能测试最终报告
* 优化整体框架，搭建ubuntu测试磁盘性能web服务器（mongdb+Apache2+vue+python+shell），以实现动态的一键测试及报告产出
* 折线图（makelines.py）中num数组生成方式
* ~~indexDemo.html中的数据来源~~
* ~~fio中磁盘参数动态获取~~
* 深入了解Locust机制
* requests 库性能不佳，打算更改测试工具为Boomer（Golang）
* Locust报告模块优化
* ~~fio测试脚本向下兼容（目前仅支持3.16版本的fio）~~
* fio脚本\`命令定义在配置文件或头部
* 整合全部类型的fio脚本
* ~~vdbench脚本用运行Python~~
* 增加PC端自动化测试框架（采用pyautogui或pywinauto待定）
* epbf调研与基础应用以分析性能瓶颈
* fio联机测试调研
* 实例一个测试流程
* ~~多个Log类实例化导致日志重复输出的问题~~
* 基本实现版本0.0.4功能，计划如下
* ~~1） 单用户，利用pymsql持续往单库单表增加数据~~
* 2） ~~单用户，多库多表增加数据，且实现测试场景写入配置文件，并校对库数据正确性~~
* 3） ~~单用户，多库多表增删改查~~，且实现容量校对及~~时间记录~~
* 4） ~~多用户，定制化压力测试~~，操作耗时折线图制作
* 5） ~~多进程、线程、用户~~，定制化测试及报告生成
* 6） 实现多客户端测试
* ~~log模块重写~~
* ~~配置模块增加~~




### 异常处理

* 现象：fio读写提示：fio: failed allocating random map. If running a large number of jobs, try the 'norandommap' option or set 'softrandommap'. Or give a larger --alloc-size to fio.
fio参数中增加：norandommap

* 现象：运行matplotlib提示：matplotlib display text must have all code points < 128 or use Unicode strings
追加`import sys
reload(sys)
sys.setdefaultencoding('utf-8')`

Common/111.py
import pymysql, time
from Common.zDecorator import tryer, timer
from Config.currency import currencyLog

_zMySqlLog = currencyLog
# 默认配置参数
default_prefix = "z_mysql_performance_"
default_db_name = "default_db"
default_table_name = "default_table"
default_db_sql = "CREATE DATABASE IF NOT EXISTS "
default_table_sql = "CREATE TABLE IF NOT EXISTS "
default_table_value = "(z_id INT UNSIGNED AUTO_INCREMENT,remarks VARCHAR(100),enter_date DATE,PRIMARY KEY ( z_id ))ENGINE=InnoDB DEFAULT CHARSET=utf8;"
default_table_insert = "insert into {} (enter_date,remarks) values (%s,%s)"
default_table_delete = "delete from {} where z_id=%s"
default_table_update = "update {} set enter_date=\"%s\",remarks=\"%s\" where z_id=%s"
default_table_select = "select count(*) from {}"

config = {'host': 'localhost',
		  'port': 3306,
		  'user': 'root',
		  'passwd': 'P@ssw0rd',
		  'charset': 'utf8'
		  # 'db' : '数据库名'
		  }


# 库类
class zMySqlDataBase():

	def __init__(self, user_config: dict, table_num: int = 1, db_name: str = None, *args, **kwargs):
		# 日志
		self.log = _zMySqlLog
		# 数据库名称
		if db_name:
			self.db_name = db_name
		else:
			self.db_name = default_db_name
		# 数据库前缀
		if 'prefix' in kwargs:
			self.prefix = kwargs["prefix"]
		else:
			self.prefix = default_prefix
		# 查询数据库，不存在则创建
		self.db_sql = default_db_sql + self.prefix + self.db_name
		try:
			self.config = user_config
			self.conn = pymysql.connect(**self.config)
			self.cursor = self.conn.cursor()
			self.cursor.execute(self.db_sql)
			self.log.logger.info("cursor in mysql")
		except:
			self.log.logger.critical("connect mysql error!")
			exit(1)

	@tryer()
	def __del__(self):
		self.log.logger.info("close mysql connect !")
		self.cursor.close()
		self.conn.close()

	# 使游标进入库中，并另游标返回字典
	@tryer()
	def connect_db(self):
		self.cursor.close()
		self.conn.close()
		self.config['db'] = self.prefix + self.db_name
		self.conn = pymysql.connect(**self.config)
		self.cursor = self.conn.cursor(cursor=pymysql.cursors.DictCursor)
		self.log.logger.info("cursor in database:" + self.prefix + self.db_name)

	# 调用该函数,使游标进入数据库中，当没有指定名称时，进入default
	@tryer()
	def connect_new_db(self, new_db):
		if 'db' in self.config:
			del self.config['db']
		self.db_name = new_db
		self.db_sql = default_db_sql + self.prefix + self.db_name
		self.config = self.config
		self.conn = pymysql.connect(**self.config)
		self.cursor = self.conn.cursor()
		self.cursor.execute(self.db_sql)
		self.log.logger.info("move cursor in new database")
		self.connect_db()

	@timer(level="DEBUG")
	def commit_change(self):
		try:
			self.conn.commit()
		except:
			self.conn.rollback()


# 表类,根据游标进行表操作
class zMySqlTable():
	def __init__(self, cursor: classmethod, table_name: str = None, *args, **kwargs):
		# 日志
		self._log = _zMySqlLog
		# 通用sql语句，用于提交，非增删改查专用
		self.a_sql = ""
		self.one = {}
		self.more = {}
		# 表名
		if table_name:
			self.table_name = table_name
		else:
			self.table_name = default_table_name
		# 不建议修改的变量
		if 'prefix' in kwargs:
			self.table_prefix = kwargs["prefix"]
		elif 'table_sql' in kwargs:
			self.table_prefix = kwargs["table_sql"]
		elif 'table_value' in kwargs:
			self.table_prefix = kwargs["table_value"]
		elif 'insert' in kwargs:
			self.table_prefix = kwargs["insert"]
		elif 'delete' in kwargs:
			self.table_prefix = kwargs["delete"]
		elif 'update' in kwargs:
			self.table_prefix = kwargs["update"]
		elif 'select' in kwargs:
			self.table_prefix = kwargs["selec"]
		else:
			self.table_prefix = default_prefix
			self.table_sql = default_table_sql
			self.table_value = default_table_value
			self.table_insert = default_table_insert.format(self.table_prefix + self.table_name)
			self.table_delete = default_table_delete.format(self.table_prefix + self.table_name)
			self.table_update = default_table_update.format(self.table_prefix + self.table_name)
			self.table_select = default_table_select.format(self.table_prefix + self.table_name)

		# 检查表是否存在sql语句，不存在则创建
		self.check_sql = self.table_sql + self.table_prefix + self.table_name + self.table_value

		# 定义游标
		self.cursor = cursor

		try:
			self.cursor.execute(self.check_sql)
		except:
			self._log.logger.error("check table error!")
			exit(1)

	@timer(level="DEBUG")
	def do_insert(self, data_options: tuple):
		self._log.logger.debug("do insert,sql=" + self.table_insert)
		self.cursor.execute(self.table_insert, data_options)

	@timer(level="DEBUG")
	def do_insert_many(self, data_options: tuple):
		self.cursor.executemany(self.table_insert, data_options)

	@timer(level="DEBUG")
	def do_delete(self, data_options: tuple):
		self._log.logger.debug("do delete,sql=" + self.table_delete % data_options)
		self.cursor.execute(self.table_delete % data_options)

	@timer(level="DEBUG")
	def do_update(self, data_options: tuple):
		self._log.logger.debug("do update,sql=" + self.table_update % data_options)
		self.cursor.execute(self.table_update % data_options)

	@timer(level="DEBUG")
	def do_select(self):
		self._log.logger.debug("do select,sql=" + self.table_select)
		self.cursor.execute(self.table_select)

	@timer(level="DEBUG")
	def do_one(self, data_options: tuple):
		self._log.logger.debug("do one，sql=" + self.a_sql % data_options)
		self.cursor(self.a_sql % data_options)

	@timer(level="DEBUG")
	def do_many(self, data_options: tuple):
		self._log.logger.debug("do many,sql=" + self.a_sql % data_options)
		self.cursor.executemany(self.a_sql % data_options)

	@timer(level="DEBUG")
	def show_one(self):
		self.one = self.cursor.fetchone()
		return self.one

	@timer(level="DEBUG")
	def show_more(self):
		self.more = self.cursor.fetchall()
		return self.more


if __name__ == "__main__":
	z1 = zMySqlDataBase(config)

	z1.connect_new_db("db1")

	t1 = zMySqlTable(cursor=z1.cursor)

	for i in range(0, 3):
		t1.do_insert_many((('2022-05-19 11:35:02', i), ('2022-05-22 11:35:02', i)))
	z1.commit_change()

	t1.do_update(('2009-05-01 11:35:02','test update zzzzz','3'))
	z1.commit_change()

	t1.do_delete('12')
	z1.commit_change()

	t1.table_select="select * from {}".format(t1.table_prefix+t1.table_name)
	t1.do_select()
	t1.show_more()
	#print(t1.more)

	del z1
# print(z1.db_name)
